nomad/state/state_store.go

```diff
@@ -1058,16 +1058,15 @@ func deleteNodeCSIPlugins(txn *memdb.Txn, node *structs.Node, index uint64) erro
 
 func deleteNodeFromPlugin(txn *memdb.Txn, plug *structs.CSIPlugin, node *structs.Node, index uint64) error {
 	plug.DeleteNode(node.ID)
+	return deleteFromPlugin(index, txn, plug)
+}
+
+// deleteFromPlugin updates a plugin but will delete it if the plugin is empty
+func deleteFromPlugin(index uint64, txn *memdb.Txn, plug *structs.CSIPlugin) error {
```

Comment:
Pulling this out into its own function split out from `deleteNodeFromPlugin` makes the Insert case "orphaned" from its purpose. This function could arguably be called `deleteEmptyPlugin` but we have this other side-effect of updating the things the caller did. It might be worth pulling the Insert case back out and into the caller.

Reply:
What about just `updateOrGCPlugin`? I like keeping the insert or delete behavior together because of that little bit of common logic.

Reply:
Sure, that works.